src/ligero/mod.rs

```diff
@@ -0,0 +1,323 @@
+use core::{borrow::Borrow, marker::PhantomData};
+
+use ark_crypto_primitives::{
+    merkle_tree::{Config, LeafParam, Path, TwoToOneParam},
+    sponge::{Absorb, CryptographicSponge},
+};
+use ark_ff::PrimeField;
+use ark_poly::Polynomial;
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use digest::Digest;
+
+use crate::{
+    Error, PCCommitment, PCCommitterKey, PCPreparedCommitment, PCPreparedVerifierKey, PCRandomness,
+    PCUniversalParams, PCVerifierKey, PolynomialCommitment,
+};
+
+use ark_std::rand::RngCore;
+
+// TODO: Disclaimer: no hiding prop
+/// The Ligero polynomial commitment scheme.
+pub struct Ligero<
+    F: PrimeField,
+    C: Config,
+    D: Digest,
+    S: CryptographicSponge,
+    /// one over the rate rho
+    const rho_inv: usize,
+    /// security parameter, used in calculating t
+    const sec_param: usize,
+> {
+    _field: PhantomData<F>,
+    _config: PhantomData<C>,
+    _digest: PhantomData<D>,
+    _sponge: PhantomData<S>,
+}
+
+// TODO come up with reasonable defaults
+const DEFAULT_RHO_INV: usize = 2;
+const DEFAULT_SEC_PARAM: usize = 128;
+
+fn calculate_t(rho_inv: usize, sec_param: usize) -> usize {
+    // TODO calculate t somehow
+    let t = 5;
+    t
+}
+
+impl<F, C, D, S, const rho_inv: usize, const sec_param: usize>
+    Ligero<F, C, D, S, rho_inv, sec_param>
+where
+    F: PrimeField + Borrow<<C as Config>::Leaf> + Absorb,
+    C: Config,
+    C::InnerDigest: Absorb,
+    D: Digest,
+    S: CryptographicSponge,
+{
+    /// Create a new instance of Ligero.
+    /// If either or both parameters are None, their default values are used.
+    pub fn new() -> Self {
+        Self {
+            _config: PhantomData,
+            _field: PhantomData,
+            _digest: PhantomData,
+            _sponge: PhantomData,
+        }
+    }
+
+    /// The verifier can check the well-formedness of the commitment by taking random linear combinations.
+    fn verify(
+        commitment: &LigeroPCCommitment<F, C>,
+        leaf_hash_params: &LeafParam<C>,
+        two_to_one_params: &TwoToOneParam<C>,
+        sponge: &mut S,
+    ) -> Result<(), Error> {
+        let t = calculate_t(rho_inv, sec_param);
+
+        // 1. Hash the received columns, to get the leaf hashes
+        let col_hashes: Vec<F> = vec![F::zero(); t];
+        for c in commitment.transcript.columns.iter() {
+            // TODO some hashing, with the digest?
```

Comment:
I think we need 1 loop over zip(columns, column_hashes, column_tree_proofs) where in each iteration we check that the column hash to the claimed hash and also verify the Merkle-proof for that hash. Maybe it even makes sense to store that info as an array of triples instead of three separate arrays.

Reply:
Sounds good, yeah the triples also crossed my mind at some point.